scripts/odbpy/io_place.py

```diff
@@ -35,52 +35,75 @@ def grid_to_tracks(origin, count, step):
     return tracks
 
 
-def equally_spaced_sequence(pin_count, possible_locations):
+def equally_spaced_sequence(side, side_pin_placement, possible_locations):
+    virtual_pin_count = 0
+    actual_pin_count = len(side_pin_placement)
+    total_pin_count = actual_pin_count + virtual_pin_count
+    for i in range(len(side_pin_placement)):
+        if isinstance(side_pin_placement[i], int): # This is an int value indicating virtual pins
+            virtual_pin_count = virtual_pin_count + side_pin_placement[i] 
+            actual_pin_count = actual_pin_count - 1 # Decrement actual pin count, this value was only there to indicate virtual pin count
+            total_pin_count = actual_pin_count + virtual_pin_count
     result = []
     tracks = len(possible_locations)
 
-    if pin_count > tracks:
-        raise Exception(
-            "There are more pins than places to put them. Try making your floorplan area larger."
-        )
-    elif pin_count == tracks:
+    if total_pin_count > tracks:
+        print(f"There are more pins/virtual_pins: {total_pin_count}, than places to put them: {tracks}. Try making your floorplan area larger.")
+        sys.exit(1)
+    elif total_pin_count == tracks:
         return possible_locations  # All positions.
-    elif pin_count == 0:
+    elif total_pin_count == 0:
         return []
 
     # From this point, pin_count always < tracks.
-    tracks_per_pin = math.floor(tracks / pin_count)  # >=1
-
+    tracks_per_pin = math.floor(tracks / total_pin_count)  # >=1
     # O| | | O| | | O| | |
     # tracks_per_pin = 3
     # notice the last two tracks are unused
     # thus:
-    used_tracks = tracks_per_pin * (pin_count - 1) + 1
+    used_tracks = tracks_per_pin * (total_pin_count - 1) + 1
     unused_tracks = tracks - used_tracks
 
     # Place the pins at those tracks...
     current_track = unused_tracks // 2  # So that the tracks used are centered
-    for _ in range(0, pin_count):
-        result.append(possible_locations[current_track])
-        current_track += tracks_per_pin
-
-    VISUALIZE_PLACEMENT = False
+    starting_track_index = current_track 
+    if virtual_pin_count == 0: # No virtual pins
+        for _ in range(0, total_pin_count):
+            result.append(possible_locations[current_track])
+            current_track += tracks_per_pin
+    else: # There are virtual pins
+        for i in range(len(side_pin_placement)):
+            if not isinstance(side_pin_placement[i], int): # We have an actual pin
+               result.append(possible_locations[current_track]) 
+               current_track += tracks_per_pin
+            else: # Virtual Pins, so just leave their needed spaces 
+                current_track += tracks_per_pin * side_pin_placement[i]
+        side_pin_placement = [pin for pin in side_pin_placement if not isinstance(pin, int)] # Remove the virtual pins from the side_pin_placement list
+    VISUALIZE_PLACEMENT = True
```

Comment:
Is there any particular reason why you changed the VISUALIZE_PLACEMENT block into something completely different instead of simply printing statistics in a separate block?

Reply:
Now the statistics are printed outside the block as you requested, I just kept the VISUALIZE_PLACEMENT = True to print out the used_track_indices which is important for understanding how the virtual pins actually works.